File: examples/python/src/shape.py
"""Abstract shape base with enum colors and the registry decorator."""

from __future__ import annotations

from abc import ABC, abstractmethod
from enum import Enum, auto
from typing import Type

from vector3 import Vector3


class Color(Enum):
    RED = auto()
    GREEN = auto()
    BLUE = auto()
    YELLOW = auto()
    WHITE = auto()


# --- Class registry via decorator ---

_shape_registry: dict[str, Type[Shape]] = {}


def register_shape(cls: Type[Shape]) -> Type[Shape]:
    """Class decorator that registers a Shape subclass by name."""
    _shape_registry[cls.__name__] = cls
    return cls


def get_registered_shapes() -> dict[str, Type[Shape]]:
    return dict(_shape_registry)


class Shape(ABC):
    def __init__(self, name: str, color: Color) -> None:
        self._name = name
        self._color = color

    @property
    def name(self) -> str:
        return self._name

    @property
    def color(self) -> Color:
        return self._color

    @abstractmethod
    def area(self) -> float: ...

    @abstractmethod
    def perimeter(self) -> float: ...

    @abstractmethod
    def centroid(self) -> Vector3[float]: ...

    def describe(self) -> str:
        return (
            f"{self.name} [{self.color.name}]  "
            f"area={self.area():.4f}  "
            f"perimeter={self.perimeter():.4f}  "
            f"centroid={self.centroid()}"
        )

    def __repr__(self) -> str:
        return f"<{self.__class__.__name__} color={self.color.name}>"

File: examples/javascript/src/shape.js
/** Color enum via frozen object with Symbol values. */
export const Color = Object.freeze({
  RED:    Symbol("RED"),
  GREEN:  Symbol("GREEN"),
  BLUE:   Symbol("BLUE"),
  YELLOW: Symbol("YELLOW"),
  WHITE:  Symbol("WHITE"),
});

export function colorName(color) {
  return Object.keys(Color).find((k) => Color[k] === color) ?? "UNKNOWN";
}

// --- Class registry ---
const _registry = new Map();

/** Class decorator (stage-3) — registers a Shape subclass by name. */
export function registerShape(cls) {
  _registry.set(cls.name, cls);
  return cls;
}

export function getRegisteredShapes() {
  return new Map(_registry);
}

/** Abstract base for all shapes. */
export class Shape {
  #name;
  #color;

  constructor(name, color = Color.BLUE) {
    if (new.target === Shape) {
      throw new TypeError("Shape is abstract — instantiate a subclass");
    }
    this.#name  = name;
    this.#color = color;
  }

  get name()  { return this.#name; }
  get color() { return this.#color; }

  area()      { throw new Error("area() not implemented"); }
  perimeter() { throw new Error("perimeter() not implemented"); }
  centroid()  { throw new Error("centroid() not implemented"); }

  describe() {
    return (
      `${this.name} [${colorName(this.color)}]  ` +
      `area=${this.area().toFixed(4)}  ` +
      `perimeter=${this.perimeter().toFixed(4)}  ` +
      `centroid=${this.centroid()}`
    );
  }

  toString() {
    return `<${this.constructor.name} color=${colorName(this.color)}>`;
  }
}
